### ISSUE TITLE
Falcon signature verification inside Miden VM

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#951]
The Falcon signature verification procedure takes a as input:
1. A public key $\mathsf{pk} := h$ where $h$ is a polynomial in $\mathbb{Z}_q[x]/(\phi)$ where $q = 1024 * 12 + 1$ and $\phi = x^{512} + 1$.
2. A message $\mathsf{m}$ 
3. A signature $\mathsf{sig} := (\mathsf{r}, \mathsf{s})$ where $\mathsf{r}$ (called "the salt") is used to hash $(\mathsf{r} || \mathsf{m})$ to a polynomial $c\in\mathbb{Z}_q[x]/(\phi)$. $\mathsf{s}$ is decoded into a polynomial $s_2 \in \mathbb{Z}[x]/(\phi)$. Note that we can interpret $s_2 \in \mathbb{Z}_q[x]/(\phi)$ but at the cost of needing to normalize its coefficients when computing the square norm later on.

The signature is valid if and only if:
1. $s_1 := c - s_2\cdot h$ in $\mathbb{Z}_q[x]/(\phi)$
2. $\lVert (s_1, s_2) \rVert^2 \leq K$ where $K$ is a fixed constant of the protocol.

# Computing $s_2\cdot h$
The most costly step of the signature verification is the polynomial product $s_2\cdot h$ in $\mathbb{Z}_q[x]/(\phi)$. This can be done in $n \log(n)$ where $n := 512$, instead of quadratically using a school-book multiplication approach, and this is the approach currently used in Miden VM.
The following note is a sketch of an algorithm that can do better, at least asymptotically, in the context of Miden VM.
To start, let's try to solve first the problem of polynomial multiplication in $\mathbb{Z}_q[x]$. To this end, let $a, b \in\mathbb{Z}_q[x]$ be two polynomials of degree less than $n$. Let $c := a\cdot b$ be the product of $a$ and $b$ which is a polynomial in $\mathbb{Z}_q[x]$ of degree less than $2n$. 
The first observation is that the Falcon prime $q$ is much smaller than the Miden VM native prime $Q := 2^{64} - 2^{32} +1$. This means that we can perform the multiplication $a\cdot b$ using the native prime $Q$ without the risk of overflow i.e. we can postpone the modular reduction til the end. In other words:

$$0 \leq a_i b_i < q \Rightarrow 0 \leq c_i < n \cdot q^2 << Q$$

The second observation is that $c$ can be provided non-deterministically and we need only verify that $a \cdot b == c$ over $\mathbb{Z}_Q[x]$ for $a, b, c \in\mathbb{Z}_q[x]$. To do this, we can use the Schwartz-Zippel lemma as follows:
1. On input $a, b \in\mathbb{Z}_q[x]$ of degree less than $n$, guess non-deterministically $c \in\mathbb{Z}_q[x]$ of degree less than $2n$.
2. Pick a random challenge $\tau \in\mathbb{Z}_Q$.
3. Evaluate $\alpha := \sum\limits_0^{n-1} a_i \tau^i$, $\beta := \sum\limits_0^{n-1} b_i \tau^i$ and $\gamma := \sum\limits_0^{2n-1} c_i \tau^i$
4. Return $1$ if $\alpha\cdot\beta = \gamma$, else return $0$.

The above protocol has soundness error $~ 2^{-54}$. In order to get soundness error smaller than $~ 2^{-100}$, we can draw the challenge $\tau$ from a quadratic extension field.
To finish, we need to reduce modulo $x^n + 1$. This reduction step yields a polynomial of degree less than $n$ given by $$\sum\limits_{i=0}^{n-1} (c_i + (- c_{n + i} \textit{ mod q}) \textit{ mod q}) x^i$$

# Asserting $\lVert (s_1, s_2) \rVert^2 \leq K$
Having computed $s_1$, the signature verification is now reduced to asserting that $\lVert (s_1, s_2) \rVert^2 \leq K$. This is achieved by normalizing the coefficients of $s_1$ and $s_2$ to be in the range $[\frac{-q}{2}, \frac{q}{2})$ and then summing the squares of the resulting coefficients. The verification concludes by asserting that this sum is less than or equal to $K$.

# Implementation
The [following](https://github.com/ASDiscreteMathematics/Miden_Lattices/blob/592873b7bcfb02f213b9d295b47773401eb50bf4/Code/falcon/falcon.py#L371) is a PoC implementation of the above ideas. Note that in the context of Miden VM, the challenge $\tau$ should be generated a-la Fiat-Shamir using the hash of polynomials $a$, $b$ and $c$.

**Comment by bobbinth** [comment]
Great write up! Thank you!

For evaluating $\alpha := \sum\limits_0^{n-1} a_i \tau^i$, $\beta := \sum\limits_0^{n-1} b_i \tau^i$ and $\gamma := \sum\limits_0^{2n-1} c_i \tau^i$, I'm wondering if we could use the following process:

First, load each polynomial from the advice provider using `adv_pipe` instruction. While we do this, we also:
  a. Compute $hash(a || b || c)$ to derive $\tau$.
  b. Compute $hash(a)$ (assuming $a$ represents the public key) and compare it to the hash of public key we use to verify the signature.

Assuming each p$a$ and $b$ have $512$ coefficients each, and $c$ has $1024$ coefficients, we'd need a total of $512$ words to store them in memory, and $64$ invocations of `adv_pipe` instruction to load them. Including hashing and other instructions, I think this step should not take much more than 300 cycles or so.

Then, compute all powers of $\tau$ up to $1024$ and save them to memory. Without specialized instructions, this will probably take around 10K cycles.

Lastly, compute evaluations using the [combine_main](https://github.com/0xPolygonMiden/miden-vm/blob/next/stdlib/asm/crypto/stark/deep_queries.masm#L4) instruction which we use for computing random linear combination of trace columns. This instruction actually does more than we need, but I think it can still work if we do the following:
- Use $(p_0, p_1)$ to aggregate the evaluation result. We don't really care about what will happen with $(r_0, r_1)$ - so, we can just ignore it.
- For OOD values (pointed to by `z_addr`) just use zeros (we may need to initialize some region of memory to zeros, and this may take up 2K - 3K cycles).
- For alpha values (pointed to by `a_addr`) use powers of $\tau$ computed in the previous step. This could take another 2K - 3K cycles.

Overall, for verifying that $a \cdot b == c$ over $\mathbb{Z}_Q[x]$ we may be able to fit into about 20K cycles.

I didn't try to estimate costs of other steps, but hopefully we can do the whole signature verification in under 50K cycles, with a stretch goal being under 32K cycles. What do you think?

**Comment by Al-Kindi-0** [comment]
Makes sense!
I am still not sure about the API though. What is provided through the stack vs the advice provider?

**Comment by bobbinth** [comment]
I haven't thought about it too much - but maybe something like this:

```
Stack input:  [MSG, PK, ...]
Stack output: [...]
```
In the above:
- `MSG` is a word representing hash of the message.
- `PK` is a word representing hash of the public key $h$.

Via the advice provider we read:
1. Preimage of `PK`.
2. Signature $(r, s)$.

I don't think we need to provide preimage of `MSG` as we assume the user verified already that the message hashes to `MSG`.

One thing that is not clear to me is how $hash(r||m)$ gets converted to polynomial $c$ and whether there is something that we'd need to provide nondeterministically to simplify this procedure.

**Comment by Al-Kindi-0** [comment]
> I don't think we need to provide preimage of `MSG` as we assume the user verified already that the message hashes to `MSG`.

I agree, the hash of the message should be the message we are signing.

> One thing that is not clear to me is how hash(r||m) gets converted to polynomial c and whether there is something that we'd need to provide nondeterministically to simplify this procedure.

This might not be straightforward. The [reference implementation](https://github.com/ASDiscreteMathematics/Miden_Lattices/blob/592873b7bcfb02f213b9d295b47773401eb50bf4/Code/falcon/falcon.py#L263) uses SHAKE256, so one thing we can think about is using RPO here. Generating the coefficients of the (hash) polynomial $c$ corresponding to the hash of $(\mathsf{r} || \mathsf{m})$ would require some bit-fiddling but shouldn't be prohibitive.

**Comment by bobbinth** [comment]
> The [reference implementation](https://github.com/ASDiscreteMathematics/Miden_Lattices/blob/592873b7bcfb02f213b9d295b47773401eb50bf4/Code/falcon/falcon.py#L263) uses SHAKE256, so one thing we can think about is using RPO here.

Yes. We can also hash the input using RPO and then do one SHAKE256 hash to be somewhat compliant - though, this would probably add around 14K cycles.


>  Generating the coefficients of the (hash) polynomial $c$ corresponding to the hash of $(\mathsf{r} || \mathsf{m})$ would require some bit-fiddling but shouldn't be prohibitive. 

What's the overall process here (at the high level)?

**Comment by Al-Kindi-0** [comment]
This is the Python code for generating $c$. The bit manipulations are contained in the `while` loop:

```python
def hash_to_point(self, message, salt):
        """
        Hash a message to a point in Z[x] mod(Phi, q).
        Inspired by the Parse function from NewHope.
        """
        n = self.n
        if q > (1 << 16):
            raise ValueError("The modulus is too large")

        k = (1 << 16) // q
        # Create a SHAKE object and hash the salt and message.
        shake = SHAKE256.new()
        shake.update(salt)
        shake.update(message)
        # Output pseudorandom bytes and map them to coefficients.
        hashed = [0 for i in range(n)]
        i = 0
        j = 0
        while i < n:
            # Takes 2 bytes, transform them in a 16 bits integer
            twobytes = shake.read(2)
            elt = (twobytes[0] << 8) + twobytes[1]  # This breaks in Python 2.x
            # Implicit rejection sampling
            if elt < k * q:
                hashed[i] = elt % q
                i += 1
            j += 1
        return hashed
```

**Comment by bobbinth** [comment]
Ah! So, SHAKE is used here to generate all coefficients. Then we should definitely start with RPO for this.

**Comment by Al-Kindi-0** [comment]
Makes sense. We can start with RPO and then take it from there.

**Comment by bobbinth** [comment]
Closed by #1000 and #1094.
